server/src/main/java/org/elasticsearch/index/shard/DenseVectorStats.java

```diff
@@ -9,73 +9,184 @@
+            out.writeBoolean(false);
+        }
+    }
+
+    static void writeFieldStatsMap(StreamOutput out, Map<String, Long> map) throws IOException {
+        out.writeMap(map, StreamOutput::writeString, StreamOutput::writeLong);
     }
 
     public void add(DenseVectorStats other) {
         if (other == null) {
             return;
         }
         this.valueCount += other.valueCount;
+        if (other.offHeapStats != null) {
+            if (this.offHeapStats == null) {
+                this.offHeapStats = other.offHeapStats;
+            } else {
+                this.offHeapStats = Stream.of(this.offHeapStats, other.offHeapStats)
+                    .flatMap(map -> map.entrySet().stream())
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, OffHeapByteSizeUtils::mergeOffHeapByteSizeMaps));
+            }
+        }
     }
 
-    /**
-     * Returns the total number of dense vectors added in the index.
-     */
+    /** Returns the total number of dense vectors added in the index. */
     public long getValueCount() {
         return valueCount;
     }
 
+    /** Returns a map of per-field off-heap stats. */
+    public Map<String, Map<String, Long>> offHeapStats() {
+        return offHeapStats;
+    }
+
+    private Map<String, Long> getTotalsByCategory() {
+        if (offHeapStats == null) {
+            return Map.of("veb", 0L, "vec", 0L, "veq", 0L, "vex", 0L);
+        } else {
+            return offHeapStats.entrySet()
+                .stream()
+                .flatMap(map -> map.getValue().entrySet().stream())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Long::sum));
+        }
+    }
+
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject(Fields.NAME);
         builder.field(Fields.VALUE_COUNT, valueCount);
+        if (params.paramAsBoolean(INCLUDE_OFF_HEAP, false)) {
+            toXContentWithFields(builder, params);
+        }
         builder.endObject();
         return builder;
     }
 
+    private void toXContentWithFields(XContentBuilder builder, Params params) throws IOException {
+        var totals = getTotalsByCategory();
+        builder.startObject(Fields.OFF_HEAP_SIZE_BYTES);
+        builder.field(Fields.TOTAL_SIZE_BYTES, totals.values().stream().mapToLong(Long::longValue).sum());
+        builder.field(Fields.TOTAL_VEB_SIZE_BYTES, totals.getOrDefault("veb", 0L));
+        builder.field(Fields.TOTAL_VEC_SIZE_BYTES, totals.getOrDefault("vec", 0L));
+        builder.field(Fields.TOTAL_VEQ_SIZE_BYTES, totals.getOrDefault("veq", 0L));
+        builder.field(Fields.TOTAL_VEX_SIZE_BYTES, totals.getOrDefault("vex", 0L));
+        if (params.paramAsBoolean(INCLUDE_PER_FIELD_STATS, false) && offHeapStats != null && offHeapStats.size() > 0) {
+            toXContentWithPerFieldStats(builder);
+        }
+        builder.endObject();
+    }
```

Comment:
we shoudl support human readable options here. I think its called `humanReadableField` or something, it will dynamically do the right thing, though maybe you need two field options (_bytes, vs. regular).

Reply:
good idea. Added [55d09e6](https://github.com/elastic/elasticsearch/pull/126704/commits/55d09e64c78353373ad5500aad401433a55d6afe)